File: 976/largest-perimeter-triangle.py
# 如果能組出最大周長的三角形，把周長（3個邊長加起來）return
# 三角形3個邊的特質：a,b,c
# a+b>c 兩邊相加，大於第3邊
# 但無法暴力試 10^4的3次方，就超過了。所以不能這樣做。
# 技巧：就從大到小，反過來查看
class Solution:
    def largestPerimeter(self, nums: List[int]) -> int:
        nums.sort()
        # print(nums)
        for i in range(len(nums)-1,1,-1):
            # 就只看最大的連續3個邊，因為不成三角形的話，跳到更小也不行
            if nums[i]<nums[i-1]+nums[i-2]:
                return nums[i]+nums[i-1]+nums[i-2]

        return 0

File: 976/largest-perimeter-triangle.cpp
// LeetCode 976. Largest Perimeter Triangle
// 要用 nums 裡的棒子長度, 組合出「三角形」兩邊和>第3邊
class Solution {
public:
    int largestPerimeter(vector<int>& nums) {
        sort(nums.begin(), nums.end()); // 排序(又快又好)
        //for (int i=nums.size()-1; i>=0; i--) { // 用迴圈, 到大小都試一次
        // 倒過來的迴圈
        //    cout << nums[i] << " "; // 印出大到小的數
        //} // 先印出「大到小」等一下,這3行會刪掉 (可留下上面3行註解, 因為可以理解「反過來的迴圈」)
        for (int i=nums.size()-1; i>=2; i--) { // 迴圈有稍修改, nums[i] vs. -1,-2
            if (nums[i] < nums[i-1] + nums[i-2]) return nums[i]+nums[i-1]+nums[i-2];
            // 如果順利「兩邊和大於第三邊」 就把我們的周長加起來, 當成答案
        }

        return 0; // 找不到任何「合法」的三角形,就 return 0
    }
};
